refactor!: optimizing `AztecNode::getLogsByTags` for new log sync algo

benesjan opened the pull request:
Fixes https://linear.app/aztec-labs/issue/F-229/improve-log-sync-performance

The return type of `getLogsByTags` did not contain block timestamp which lead to the new log sync algo being inefficient - we needed to perform a lot of calls to `getBlockHeaderByHash` endpoint to get the timestamp.

In this PR I include the timestamp in the return value.

This is one of 3 PRs in which I clean up the `getLogsByTags` endpoint:
1. This is the first PR - including block timestamp,
2. in the followup PR I will type the tag arg of the function to be `SiloedTag`,
3. in the last PR I will drop pagination from this endpoint.

Re point 3 - pagination here doesn't really make sense because we get more logs per tag only if there are multiple devices sending logs from a given sender to a recipient.

benesjan commented:
* **#19290** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19290?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19289** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19289?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19233** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19233?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19125** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19125?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19161** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19161?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19158** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19158?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19157** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19157?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19157?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#19030** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19030?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17611** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17611?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17568** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17568?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17534** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17534?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17445** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17445?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17451** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17451?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

spalladino commented on yarn-project/stdlib/src/types/shared.ts:
```diff
@@ -30,6 +32,12 @@ export type UInt32 = number;
+ * Zod schema for UInt64 that accepts bigint, number, or string (from JSON) and coerces to bigint.
+ * Inspired by BlockNumberSchema.
+ */
+export const UInt64Schema = z.union([z.bigint(), z.number(), z.string()]).pipe(z.coerce.bigint());
```
Nit: you probably want to add this in `foundation/src/schemas/schemas.ts`

benesjan replied:
Was not aware of that. Moved it in https://github.com/AztecProtocol/aztec-packages/pull/19157/commits/46d83bcd42a0a28b118c108d9745dd0c038a0e58

Thanks!

benesjan committed:
refactor!: optimizing `AztecNode::getLogsByTags` for new log sync algo

Fixes https://linear.app/aztec-labs/issue/F-229/improve-log-sync-performance

The return type of `getLogsByTags` did not contain block timestamp which lead to the new log sync algo being inefficient - we needed to perform a lot of calls to `getBlockHeaderByHash` endpoint to get the timestamp.

In this PR I include the timestamp in the return value.

This is one of 3 PRs in which I clean up the `getLogsByTags` endpoint:
1. This is the first PR - including block timestamp,
2. in the followup PR I will type the tag arg of the function to be `SiloedTag`,
3. in the last PR I will drop pagination from this endpoint.

Re point 3 - pagination here doesn't really make sense because we get more logs per tag only if there are multiple devices sending logs from a given sender to a recipient.
